src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -293,3 +290,23 @@
       wrap-with-certificate-cn
       wrap-with-default-body
       wrap-with-debug-logging))
+
+(defn wrap-with-parent-check
+  "Middleware that checks the parent exists before serving the rest of the
+   application. This ensures we always return 404's on child paths when the
+   parent data is empty."
+  [app version parent id]
+  (fn [{:keys [globals] :as req}]
+    (let [{:keys [scf-read-db url-prefix]} globals
+          status-fn (case parent
```

Comment:
I just hate passing around functions if you don't have to, it feels like action-at-a-distance unless you absolutely need it. Let me ponder this again.